### ISSUE TITLE
undef on a hash slice only undefines one value

### ISSUE THREAD

**Issue by snoopyjc** [Perl/perl5#20537]
This is a bug report for perl from snoopyjc@gmail.com,
generated with the help of perlbug 1.42 running under perl 5.36.0.

-----------------------------------------------------------------
[Please describe your issue here]

    use Data::Dumper;
    my %hash = (k1=>'v1', k2=>'v2', k3=>'v3');
    my @to_undef = ('k1', 'k2');
    undef @hash{@to_undef};
    print Dumper(\%hash) . "\n";

I would expect this to undef the values associated with both k1 and k2, but here is the output:

    $VAR1 = {
          'k2' => undef,
          'k3' => 'v3',
          'k1' => 'v1'
        };

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=core
    severity=medium
---
Site configuration information for perl 5.36.0:

Configured by VssAdministrator at Wed Jul 27 10:26:14 UTC 2022.

Summary of my perl5 (revision 5 version 36 subversion 0) configuration:

  Platform:
    osname=msys
    osvers=3.3.5-341.x86_64
    archname=x86_64-msys-thread-multi
    uname='msys_nt-10.0-20348 fv-az131-318 3.3.5-341.x86_64 2022-07-08 09:41 utc x86_64 msys '
    config_args='-des -Dusethreads -Doptimize=-march=x86-64 -mtune=generic -O2 -pipe -Dprefix=/usr -Dvendorprefix=/usr -Dprivlib=/usr/share/perl5/core_perl -Darchlib=/usr/lib/perl5/core_perl -Dsitelib=/usr/share/perl5/site_perl -Dsitearch=/usr/lib/perl5/site_perl -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib/perl5/vendor_perl -Dscriptdir=/usr/bin/core_perl -Dsitescript=/usr/bin/site_perl -Dvendorscript=/usr/bin/vendor_perl -Dinc_version_list=none -Dman1ext=1perl -Dman3ext=3perl -Darchname=-msys-threads -Dmyarchname=-msys -Dlibperl=msys-perl5_36.dll -Dcc=gcc -Dld=g++ -Accflags=-march=x86-64 -mtune=generic -O2 -pipe -fwrapv'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -march=x86-64 -mtune=generic -O2 -pipe -fwrapv -fno-strict-aliasing -fstack-protector-strong -D_FORTIFY_SOURCE=2'
    optimize='-march=x86-64 -mtune=generic -O2 -pipe'
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -D_GNU_SOURCE -march=x86-64 -mtune=generic -O2 -pipe -fwrapv -fno-strict-aliasing -fstack-protector-strong'
    ccversion=''
    gccversion='11.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong'
    libpth=/usr/lib
    libs=-lpthread -ldl -lcrypt
    perllibs=-lpthread -ldl -lcrypt
    libc=/usr/lib/libmsys-2.0.a
    so=dll
    useshrplib=true
    libperl=msys-perl5_36.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong'

---
@INC for perl 5.36.0:
    /usr/lib/perl5/site_perl
    /usr/share/perl5/site_perl
    /usr/lib/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib/perl5/core_perl
    /usr/share/perl5/core_perl

---
Environment for perl 5.36.0:
    HOME=/c/Users/snoopyjc
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    MSYS=enable_pcon
    MSYSTEM=MINGW64
    MSYSTEM_CARCH=x86_64
    MSYSTEM_CHOST=x86_64-w64-mingw32
    MSYSTEM_PREFIX=/mingw64
    PATH=/c/Users/jo2742/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/bin:/usr/bin:/c/Users/snoopyjc/bin:/c/Program Files (x86)/Common Files/Oracle/Java/javapath:/c/windows/system32:/c/windows:/c/windows/System32/Wbem:/c/windows/System32/WindowsPowerShell/v1.0:/c/windows/System32/OpenSSH:/c/Program Files/1E/Client/Extensibility/NomadBranch:/cmd:/c/Users/snoopyjc/AppData/Local/Programs/Python/Python310/Scripts:/c/Users/snoopyjc/AppData/Local/Programs/Python/Python310:/c/Users/snoopyjc/AppData/Local/Microsoft/WindowsApps:/c/pypy/pypy3.9-v7.3.9-win64:/usr/bin/vendor_perl:/usr/bin/core_perl:/c/GnuWin32/bin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/bash

**Comment by Grinnz** [comment]
undef on a hash slice is not defined behavior. This should throw an error if anything.

> Use only on a scalar value, an array (using @), a hash (using %), a subroutine (using &), or a typeglob (using *).

https://perldoc.perl.org/functions/undef

**Comment by Grinnz** [comment]
(delete is defined to operate on hash slices, however, if that operation is acceptable to your use case)

**Comment by snoopyjc** [comment]
Ok - how about this - since `delete @hash{@keys}` is defined as:

    foreach my $key (@keys) {
        delete $hash{$key};
    }

How about we define `undef @hash{@keys}` as:

    foreach my $key (@keys) {
        undef $hash{$key};
    }

The current functionality as `undef $hash{$keys[-1]}` makes no sense.

I propose to update the test case and the code and submit a PR over the holidays.

**Comment by leonerd** [comment]
By analogy to `undef @arr`, `undef %hash`, and the various `delete` cases, I would casually expect that `undef @hash{@keys}` would indeed undefine all of them. I suspect the reason it doesn't is just down to an oversight, rather than any fundamental reason why it shouldn't.

**Comment by Leont** [comment]
> By analogy to undef @arr, undef %hash, and the various delete cases, I would casually expect that undef @hash{@keys} would indeed undefine all of them. I suspect the reason it doesn't is just down to an oversight, rather than any fundamental reason why it shouldn't.

Agreed. There's exactly one sensible thing to do in this case, so there's no reason for it not to do that thing.

**Comment by snoopyjc** [comment]
> > By analogy to undef @arr, undef %hash, and the various delete cases, I would casually expect that undef @hash{@keys} would indeed undefine all of them. I suspect the reason it doesn't is just down to an oversight, rather than any fundamental reason why it shouldn't.
> 
> Agreed. There's exactly one sensible thing to do in this case, so there's no reason for it not to do that thing.

How can we remove the "Closable" status?

**Comment by richardleach** [comment]
Implementation questions:
* Is the return value still a single `PL_sv_undef`?
* What's the desired behaviour when one or more of the specified keys isn't present in the hash? Does it throw a warning or stay quiet? Does it autovivify the missing keys?
* Works on array slices too?

**Comment by snoopyjc** [comment]
> Implementation questions:
> 
> * Is the return value still a single `PL_sv_undef`?
Yes
> * What's the desired behaviour when one or more of the specified keys isn't present in the hash? Does it throw a warning or stay quiet? Does it autovivify the missing keys?
Quiet, yes
> * Works on array slices too?
yes

Details - `undef` on a list of keys should be like a loop calling `undef` on each of the keys, hence quiet and autoviv:

    use Data::Dumper;
    undef $hash{key};
    print Dumper(\%hash);
    $VAR1 = {
          'key' => undef
        };

Undef on an array slice should work similarly (like a loop calling `undef` on each index):

    use Data::Dumper;
    undef $arr[1];
    print Dumper(\@arr);
    $VAR1 = [
          undef,
          undef
        ];

Here shows it NOT working on an array slice:

    use Data::Dumper;
    undef $arr[5,3,1];
    print Dumper(\@arr);
    $VAR1 = [
          undef,
          undef
        ];

**Comment by iabyn** [comment]
On Thu, Nov 24, 2022 at 09:38:33PM -0800, Joe Cool wrote:
> Undef on an array slice already works:
> 
>     use Data::Dumper;
>     undef $arr[1,3,5];
>     print ***@***.***);
>     $VAR1 = [
>           undef,
>           undef,
>           undef,
>           undef,
>           undef,
>           undef
>         ];

But not in the way you think:

    my @a = qw(a b c d e f g h);
    undef @a[1,3,5];
    use Data::Dumper;
    print Dumper ***@***.***;

    $VAR1 = [
              'a',
              'b',
              'c',
              'd',
              'e',
              undef,
              'g',
              'h'
            ];

The undef imposes scalar context on the array slice, so it just returns
the final element, which is then undeffed:

    $ p -E'my @a = qw(a b c d e f g h); my $s = @a[1,3,5]; say $s;'
    f
    $

Similarly, undef @hash{ ... keys ... } already undefs the value associated
with the last key.


-- 
I took leave and went to hear Mrs Turner's daughter play on the harpsicon,
but Lord, it was enough to make any man sick to hear her; yet I was forced
to commend her highly.
    -- The Diary of Samuel Pepys, 1 May 1663

**Comment by iabyn** [comment]
That's slightly weird. Somewhere, my backslash-arrayname's got replaced with ***@***.***.

**Comment by leonerd** [comment]
> That's slightly weird. Somewhere, my backslash-arrayname's got replaced with _**@**_.***.

I wonder if that's github "trying to protect email addresses". It turns out that Perl's use of @ for array variable names tends to collide with email, twitter, telegram, mastodon, .... It's getting more annoying lately on some of those platforms. :/

**Comment by haarg** [comment]
undef on an array is currently allowed, and will empty the array. That is not consistent with the requested behavior of undef on a list.

**Comment by snoopyjc** [comment]
> On Thu, Nov 24, 2022 at 09:38:33PM -0800, Joe Cool wrote: Undef on an array slice already works: use Data::Dumper; undef $arr[1,3,5]; print ***@***.***); $VAR1 = [ undef, undef, undef, undef, undef, undef ];
> But not in the way you think: my @A = qw(a b c d e f g h); undef @A[1,3,5]; use Data::Dumper; print Dumper ***@***.***; $VAR1 = [ 'a', 'b', 'c', 'd', 'e', undef, 'g', 'h' ]; The undef imposes scalar context on the array slice, so it just returns the final element, which is then undeffed: $ p -E'my @A = qw(a b c d e f g h); my $s = @A[1,3,5]; say $s;' f $ Similarly, undef @hash{ ... keys ... } already undefs the value associated with the last key.
> […](#)
> -- I took leave and went to hear Mrs Turner's daughter play on the harpsicon, but Lord, it was enough to make any man sick to hear her; yet I was forced to commend her highly. -- The Diary of Samuel Pepys, 1 May 1663

Yeah - I tested it again with [5,3,1] and it only undef'd 1.  I edited my comment above.
